rl4co/models/zoo/am/decoder.py

```diff
@@ -0,0 +1,210 @@
+from dataclasses import dataclass
+from typing import Tuple, Union
+
+import torch
+import torch.nn as nn
+
+from einops import rearrange
+from tensordict import TensorDict
+from torch import Tensor
+
+from rl4co.envs import RL4COEnvBase
+from rl4co.models.common.constructive.autoregressive.decoder import AutoregressiveDecoder
+from rl4co.models.nn.attention import PointerAttention
+from rl4co.models.nn.env_embeddings import env_context_embedding, env_dynamic_embedding
+from rl4co.models.nn.env_embeddings.dynamic import StaticEmbedding
+from rl4co.utils.ops import batchify, unbatchify
+from rl4co.utils.pylogger import get_pylogger
+
+log = get_pylogger(__name__)
+
+
+@dataclass
+class PrecomputedCache:
+    node_embeddings: Tensor
+    graph_context: Union[Tensor, float]
+    glimpse_key: Tensor
+    glimpse_val: Tensor
+    logit_key: Tensor
+
+
+class AttentionModelDecoder(AutoregressiveDecoder):
+    """
+
+    # TODO
+    Auto-regressive decoder for constructing solutions for combinatorial optimization problems.
+    Given the environment state and the embeddings, compute the logits and sample actions autoregressively until
+    all the environments in the batch have reached a terminal state.
+    We additionally include support for multi-starts as it is more efficient to do so in the decoder as we can
+    natively perform the attention computation.
+
+
+
+    Args:
+        # TODO
+        env_name: environment name to solve
+        embed_dim: Dimension of the embeddings
+        num_heads: Number of heads for the attention
+        use_graph_context: Whether to use the initial graph context to modify the query
+        context_embedding: Module to compute the context embedding. If None, the default is used
+        dynamic_embedding: Module to compute the dynamic embedding. If None, the default is used
+    """
+
+    def __init__(
+        self,
+        embed_dim: int = 128,
+        num_heads: int = 8,
+        env_name: Union[str, RL4COEnvBase] = "tsp",
```

Comment:
Is it necessary to support a `RL4COEnvBase` input? Will this be confusing?

Reply:
Good suggestion, I guess we can make it only as string to avoid confusion and overhead

Reply:
I will do this now